### DISCUSSION TITLE
Provide an SDK or API to facilitate programmatic operation or integration of JuiceFS

### DISCUSSION THREAD

**Discussion by shumin1027** [juicedata/juicefs#3550] [General]
At present, JuiceFS can only be operated through the `juicefs` command, which is not particularly friendly to programming. It is necessary to provide an SDK or API to facilitate programmatic operation or integration of JuiceFS

**Comment by SandyXSD** [comment]
Could you provide more details? Which subcommand and under what scenario that we can benefit from using SDK instead of executing the `juciefs` command.

**Comment by shumin1027** [comment]
> Could you provide more details? Which subcommand and under what scenario that we can benefit from using SDK instead of executing the `juciefs` command.

 If you use the `go` language to call `format`, `config`, `destroy` and other basic operations, currently you can only splice strings like "juicefs format ****" and execute commands through `exec` To complete, you cannot directly call the SDK function to operate JuiceFS,
At present, the juicefs code does not seem to organize and encapsulate such a set of low-level function interfaces. Many logics are implemented directly in the `cmd` layer, so there is no way to directly reference them.

![image](https://user-images.githubusercontent.com/5450783/235095968-9506f10c-bbb1-422b-a4ae-af50822112e9.png)

For example, can this `format` operation be encapsulated in another layer, decoupled from the `cmd` package, and provide an interface like 
```go
 func Format(args FormatArgs) error{}
```
which can be directly called by the program.


Let the `CLI` be one of the ways to operate JuiceFS, not the only way

**Comment by SandyXSD** [comment]
These operations are for management, and usually not used frequently. So CLI should be a more convenient and general way. Recently we have no plan to add SDK for them.
